pkg/mutation/system.go

```diff
@@ -85,16 +85,23 @@ func (s *System) Mutate(obj *unstructured.Unstructured, ns *corev1.Namespace) er
 			if m.Matches(obj, ns) {
 				err := m.Mutate(obj)
 				if err != nil {
-					return errors.Wrapf(err, "Mutation failed for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
+					return false, errors.Wrapf(err, "mutation failed for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
 				}
 			}
 		}
 		if cmp.Equal(old, obj) {
-			return nil
+			if i == 0 {
+				return false, nil
+			}
+			if cmp.Equal(original, obj) {
```

Comment:
Do we want to do deep comparisons to find oscillating mutations? This could ~double the cost of mutation as I imagine for large objects the deep comparison is likely the most expensive part. If we get an oscillating mutation, we could just wait for a non-convergence error.

I think the tradeoff here is: if we have a system of 20 mutations and we detect oscillation after the 2nd application, we catch it early, but successful mutations now have ~40 comparisons.

Without this check: if we have the same oscillating system, we detect it after ~20 mutations, but successful mutations now have ~20 comparisons.

Reply:
Note that if line 92 is true, we will always return, either with an error or without.
The number of mutations will stay the same as before the change, and the number of comparison will be increased by 1 only (always).
If we have 20 mutations, and end after the second iteration, both before the change we have 40 mutation, and 3 comparisons (2 with old, 1 with original).

```
		if cmp.Equal(old, obj) { // If this is true, we always leave the if with a return
			if i == 0 {
				return false, nil
			}
			if cmp.Equal(original, obj) { // If not true, we return anyway
				return false, fmt.Errorf("ocilating mutation for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
			}
			return true, nil // We always return here anyway
		} 
```

Reply:
You are right, sorry for the invalid feedback!